cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -0,0 +1,332 @@
+
+  thrust::copy(rmm::exec_policy(stream)->on(stream),
+               poly_idx.begin<uint32_t>(),
+               poly_idx.end<uint32_t>(),
+               d_poly_idx.begin());
+
+  rmm::device_uvector<uint32_t> d_quad_idx(num_pairs, stream);
+
+  thrust::copy(rmm::exec_policy(stream)->on(stream),
+               quad_idx.begin<uint32_t>(),
+               quad_idx.end<uint32_t>(),
+               d_quad_idx.begin());
+
+  // sort (d_poly_idx, d_quad_idx) using d_quad_idx as key => (quad_idxs, poly_idxs)
+  thrust::sort_by_key(
+    rmm::exec_policy(stream)->on(stream), d_quad_idx.begin(), d_quad_idx.end(), d_poly_idx.begin());
+
+  // reduce_by_key using d_quad_idx as the key
+  // exclusive_scan on numbers of polygons associated with a quadrant to create d_poly_idx_offsets
+
+  rmm::device_uvector<uint32_t> d_poly_idx_offsets(num_pairs, stream);
+
+  uint32_t num_quads =
+    thrust::distance(d_poly_idx_offsets.begin(),
+                     thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
+                                           d_quad_idx.begin(),
+                                           d_quad_idx.end(),
+                                           thrust::constant_iterator<uint32_t>(1),
+                                           d_quad_idx.begin(),
+                                           d_poly_idx_offsets.begin())
+                       .second);
+
+  d_quad_idx.resize(num_quads, stream);
+  d_poly_idx_offsets.resize(num_quads, stream);
+
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         d_poly_idx_offsets.begin(),
+                         d_poly_idx_offsets.end(),
+                         d_poly_idx_offsets.begin());
+
+  auto point_index_col   = make_fixed_width_column<uint32_t>(point_x.size(), stream, mr);
+  auto poly_index_col    = make_fixed_width_column<uint32_t>(point_x.size(), stream, mr);
+  auto poly_distance_col = make_fixed_width_column<T>(point_x.size(), stream, mr);
+
+  find_nearest_polyline_kernel<T><<<num_quads, threads_per_block, 0, stream>>>(
+    d_quad_idx.begin(),
+    d_poly_idx_offsets.size(),
+    d_poly_idx_offsets.begin(),
+    d_poly_idx.size(),
+    d_poly_idx.begin(),
+    *cudf::column_device_view::create(quad_lengths, stream),
+    *cudf::column_device_view::create(quad_offsets, stream),
+    *cudf::column_device_view::create(point_indices, stream),
+    *cudf::column_device_view::create(point_x, stream),
+    *cudf::column_device_view::create(point_y, stream),
+    *cudf::column_device_view::create(poly_offsets, stream),
+    *cudf::column_device_view::create(poly_points_x, stream),
+    *cudf::column_device_view::create(poly_points_y, stream),
+    *cudf::mutable_column_device_view::create(*point_index_col, stream),
+    *cudf::mutable_column_device_view::create(*poly_index_col, stream),
+    *cudf::mutable_column_device_view::create(*poly_distance_col, stream));
+
+  CUDA_TRY(cudaStreamSynchronize(stream));
```

Comment:
As elsewhere, probably should synchronize outside of this function, since this function takes a stream.